runtime/doc/develop.txt

```diff
@@ -159,7 +159,200 @@ VIM IS... NOT						*design-not*
+   of the word will cause the mechanism to fail.  Also, this becomes slow when
+   the bad word is quite different from the good word.
+
+The choice made is to use the second mechanism and use a separate file.  This
+way a user with sufficient memory can get very good suggestions while a user
+who is short of memory or just wants the spell checking and no suggestions
+doesn't use so much memory.
+
+
+Word frequency
+
+For sorting suggestions it helps to know which words are common.  In theory we
+could store a word frequency with the word in the dictionary.  However, this
+requires storing a count per word.  That degrades word tree compression a lot.
+And maintaining the word frequency for all languages will be a heavy task.
+Also, it would be nice to prefer words that are already in the text.  This way
+the words that appear in the specific text are preferred for suggestions.
+
+What has been implemented is to count words that have been seen during
+displaying.  A hashtable is used to quickly find the word count.  The count is
+initialized from words listed in COMMON items in the affix file, so that it
+also works when starting a new file.
+
+This isn't ideal, because the longer Vim is running the higher the counts
+become.  But in practice it is a noticeable improvement over not using the word
+count.
+
+==============================================================================
+3. Assumptions						*design-assumptions*
+
+The following sections define the portability and compatibility constraints that
+all Vim code and build tools must adhere to.
+
+
+MAKEFILES					*assumptions-makefiles*
+						*POSIX.1-2001*
+
+Our Makefiles target maximum portability, relying solely on features
+defined in POSIX.1-2001 `make` and ignoring later POSIX standards or
+GNU/BSD extensions.  In practical terms, avoid:
+
+	– % pattern rules
+	– modern assignment (`:=`, `::=`) outside POSIX.1-2001
+	– special targets (`.ONESHELL`, `.NOTPARALLEL`, `.SILENT`, …)
+	– order-only prerequisites (`|`) or automatic directory creation
+	– GNU/BSD conditionals (`ifdef`, `ifndef`, `.for`/`.endfor`, …)
+
+Since POSIX.1-2001 supports only traditional suffix rules, every object
+built in a separate directory must have an explicit rule.  For example:
+
+	objects/evalbuffer.o: evalbuffer.c
+		$(CCC) -o $@ evalbuffer.c
+
+This verbosity ensures that the same Makefile builds Vim unchanged with
+the default `make` on Linux, *BSD, macOS, Solaris, AIX, HP-UX and virtually
+any Unix-like OS.
+
+
+C COMPILER					*assumptions-C-compiler*
+						*ANSI-C* *C89* *C99*
+
+We require a mostly C99 compliant C compiler.  Because not all maintained
+compilers fully implement C99, we use only a subset and explicitly forbid
+certain features.  This will gradually be adjusted over time.
+
+Features not to be used ~
+
+These C99 features are not to be used, because MSVC 2015 (in Vim CI, extended
+support until October 14, 2025) does not support them:
```

Comment:
@chrisbra 
Does Vim support the OpenVMS C compilers listed in the following links?

https://github.com/vim/vim/pull/16962#issuecomment-2748278248
> OpenVMS version - x86_64
> Compiler: VSI C x86-64 V7.5-009 (GEM 50XBR) on OpenVMS x86_64 V9.2-2
> 
> OpenVMS version - VAX
> Compiler: Compaq C V6.4-005 on OpenVMS VAX V7.3
> 
> OpenVMS version - Alpha
> Compiler: Compaq C V6.5-001 on OpenVMS Alpha V7.3

None of the three seem to be fully C99 compliant.
The VAX version in particular is quite old, with the last release being in 2001.

Reply:
Hmm, not being able to use newer C standard features just to support these old compilers is a real drawback. It might not be the right time yet, but I believe we should consider eventually dropping support for them.